File: scripts/governor_zen_reflect.py
#!/usr/bin/env python3
"""Zen reflect-before-commit gate for CNET governor.

Sit still: re-check scoreboard + persona affect before heavy muscles fire.
Non-attachment: drop thrash goals when stuck.
Engaged: never delay keep_teacher_alive / clear infra fires.

Does not override pins or eval_veto — only softens/redirects agenda.
"""
from __future__ import annotations

import json
import os
import time
from pathlib import Path
from typing import Any

ROOT = Path(os.environ.get("CNET_ROOT", Path(__file__).resolve().parents[1]))
GOV = Path(os.environ.get("CNET_GOVERNOR_DIR", ROOT / "logs/governor"))
OUT = GOV / "zen_reflect.json"

HEAVY = {
    "run_cert_learn",
    "run_structure_mine",
    "inject_window_gaps",
    "break_plateau",
}
ALWAYS_COMMIT = {"ensure_bonsai", "ensure_lane", "close_outcome", "log_hold", "run_eval_probe"}


def reflect(
    picked: list[dict],
    sb: dict[str, Any],
    state: dict[str, Any],
    persona: dict[str, Any] | None = None,
) -> tuple[list[dict], dict[str, Any]]:
    """Return possibly revised goal list + reflect log."""
    persona = persona or {}
    affect = persona.get("affect") or sb.get("persona_affect") or {}
    traits = persona.get("traits") or sb.get("persona_traits") or {}
    vigilance = float(affect.get("vigilance") or sb.get("affect_vigilance") or 0.5)
    frustration = float(affect.get("frustration") or sb.get("affect_frustration") or 0.5)
    caution = float(traits.get("caution") or sb.get("persona_caution") or 0.5)
    patience = float(traits.get("patience") or 0.5)

    principles: list[str] = []
    actions_note: list[str] = []
    revised = list(picked)

    # 1) Infra fire — move without hesitation
    if int(sb.get("bonsai_ok") or 0) == 0 or int(sb.get("lane_ok") or 0) == 0:
        principles.append("engaged_action: teacher/lane down — commit keep_alive immediately")
        # ensure keep_teacher first if present in charter picks or inject
        ids = [g.get("id") for g in revised]
        if "keep_teacher_alive" not in ids:
            revised.insert(0, {"id": "keep_teacher_alive", "priority": 0, "actions": ["ensure_bonsai", "ensure_lane"], "when": "true"})
        log = _log(principles, actions_note, revised, "commit_infra")
        return revised, log

    # 2) Eval veto / freeze — non-attachment to seal plans
    if int(sb.get("eval_veto") or 0) or int(sb.get("freeze_seals") or 0):
        principles.append("non_attachment: eval_veto/freeze — drop seal-heavy goals")
        before = [g.get("id") for g in revised]
        revised = [
            g
            for g in revised
            if g.get("id")
            not in ("peft_jtc", "structure_mine", "break_plateau")
        ]
        if not revised:
            revised = [{"id": "outcome_review", "priority": 3, "actions": ["close_outcome", "run_eval_probe"], "when": "true"}]
        actions_note.append(f"dropped_seals:{before}->{[g.get('id') for g in revised]}")

    # 3) Attachment / thrash detection
    streak = int(state.get("same_goals_streak") or 0)
    last = state.get("last_goals") or []
    now_ids = [g.get("id") for g in revised]
    if last == now_ids:
        streak += 1
    else:
        streak = 1
    state["same_goals_streak"] = streak
    d_backlog = float(sb.get("d_backlog_pressure") or 0)
    if streak >= 4 and d_backlog >= 0 and "outcome_review" not in now_ids:
        principles.append("beginners_mind: attached to same goals without backlog improvement — force review")
        revised = [
            {"id": "outcome_review", "priority": 2, "actions": ["close_outcome", "run_eval_probe"], "when": "true"},
            {"id": "health_hold", "priority": 9, "actions": ["log_hold"], "when": "true"},
        ]
        actions_note.append("attachment_break")

    # 4) High vigilance / caution — sit still on heavy thrash
    vig_hold = 0.72
    if (vigilance >= vig_hold or caution >= 0.78) and int(sb.get("busy") or 0):
        principles.append("zazen: high vigilance+busy — demote heavy explore/break")
        revised = [
            g
            for g in revised
            if g.get("id") not in ("break_plateau", "coverage_curiosity", "verified_web")
        ]
        if not revised:
            revised = [{"id": "health_hold", "priority": 9, "actions": ["log_hold"], "when": "true"}]
        actions_note.append("vigilance_hold")

    # 5) Frustration without patience → thoroughness: prefer review over inject spam
    if frustration >= 0.65 and patience >= 0.55 and any(g.get("id") == "drain_open_gaps" for g in revised):
        if streak >= 3:
            principles.append("simplicity: inject thrash — insert outcome_review before more gaps")
            if not any(g.get("id") == "outcome_review" for g in revised):
                revised.insert(0, {"id": "outcome_review", "priority": 2, "actions": ["close_outcome", "run_eval_probe"], "when": "true"})
                revised = revised[:3]
            actions_note.append("frustration_review")

    # 6) Present moment: strip empty
    revised = [g for g in revised if g.get("id")]
    if not revised:
        principles.append("present_moment: nothing left — health_hold")
        revised = [{"id": "health_hold", "priority": 9, "actions": ["log_hold"], "when": "true"}]

    if not principles:
        principles.append("clear_see: agenda accepted — move without hesitation")

    log = _log(principles, actions_note, revised, "ok")
    return revised, log


def filter_actions(actions: list[str], sb: dict[str, Any], zen_log: dict[str, Any]) -> list[str]:
    """Optional second gate on action names."""
    if zen_log.get("mode") == "commit_infra":
        return actions
    if int(sb.get("eval_veto") or 0) or int(sb.get("freeze_seals") or 0):
        return [a for a in actions if a not in ("run_cert_learn", "run_structure_mine") or a in ALWAYS_COMMIT]
    return actions


def _log(principles: list[str], notes: list[str], goals: list[dict], mode: str) -> dict[str, Any]:
    log = {
        "ts": time.strftime("%Y-%m-%dT%H:%M:%S%z"),
        "mode": mode,
        "principles": principles,
        "notes": notes,
        "goals_after": [g.get("id") for g in goals],
        "engine": "zen_reflect_v1",
    }
    GOV.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(log, indent=2) + "\n")
    return log


def self_test() -> int:
    sb = {"bonsai_ok": 1, "lane_ok": 1, "eval_veto": 0, "freeze_seals": 0, "busy": 0, "d_backlog_pressure": 0}
    st: dict[str, Any] = {"last_goals": ["drain_open_gaps"], "same_goals_streak": 3}
    picked = [{"id": "drain_open_gaps", "actions": ["inject_window_gaps"]}]
    # attachment
    st["same_goals_streak"] = 3
    st["last_goals"] = ["drain_open_gaps"]
    r, log = reflect(picked, {**sb, "d_backlog_pressure": 1}, st, {"affect": {"vigilance": 0.4, "frustration": 0.7}, "traits": {"caution": 0.6, "patience": 0.7}})
    # after reflect streak becomes 4
    assert "outcome_review" in [g["id"] for g in r] or log.get("notes")
    # infra fire
    r2, log2 = reflect([], {"bonsai_ok": 0, "lane_ok": 1}, {}, {})
    assert r2[0]["id"] == "keep_teacher_alive"
    assert log2["mode"] == "commit_infra"
    # veto drops seals
    r3, _ = reflect(
        [{"id": "peft_jtc", "actions": ["run_cert_learn"]}],
        {**sb, "eval_veto": 1},
        {},
        {},
    )
    assert all(g["id"] != "peft_jtc" for g in r3)
    print("ZEN_REFLECT_SELFTEST_PASS checks=3")
    return 0


if __name__ == "__main__":
    import sys

    if "--test" in sys.argv:
        raise SystemExit(self_test())
    print("usage: import reflect() from governor")

File: tools/governor_zen_reflect.c
/* Zen reflect-before-commit gate for CNET governor (core).
 *
 * Usage:
 *   governor_zen_reflect --test
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#ifdef _WIN32
#include <direct.h>
#define MKDIR(p) _mkdir(p)
#else
#include <sys/stat.h>
#define MKDIR(p) mkdir((p), 0755)
#endif

#define MAX_GOALS 16
#define ID_LEN 64

typedef struct {
    char id[ID_LEN];
    char actions[256];
} Goal;

typedef struct {
    char mode[32];
    char principles[8][160];
    int n_principles;
    char notes[8][128];
    int n_notes;
    char goals_after[MAX_GOALS][ID_LEN];
    int n_goals;
} ZenLog;

static void ensure_dir(const char *path) {
    char tmp[512];
    size_t i, len;
    snprintf(tmp, sizeof tmp, "%s", path);
    len = strlen(tmp);
    for (i = 1; i < len; i++) {
        if (tmp[i] == '/' || tmp[i] == '\\') {
            char c = tmp[i];
            tmp[i] = 0;
            MKDIR(tmp);
            tmp[i] = c;
        }
    }
    MKDIR(tmp);
}

static void add_prin(ZenLog *log, const char *s) {
    if (log->n_principles < 8)
        snprintf(log->principles[log->n_principles++], 160, "%s", s);
}

static void add_note(ZenLog *log, const char *s) {
    if (log->n_notes < 8)
        snprintf(log->notes[log->n_notes++], 128, "%s", s);
}

static int has_id(Goal *g, int n, const char *id) {
    int i;
    for (i = 0; i < n; i++)
        if (strcmp(g[i].id, id) == 0) return 1;
    return 0;
}

static void write_log(const ZenLog *log) {
    const char *gov = getenv("CNET_GOVERNOR_DIR");
    char path[512];
    FILE *f;
    int i;
    if (!gov || !gov[0]) gov = "logs/governor";
    ensure_dir(gov);
    snprintf(path, sizeof path, "%s/zen_reflect.json", gov);
    f = fopen(path, "w");
    if (!f) return;
    fprintf(f, "{\n  \"mode\": \"%s\",\n  \"engine\": \"zen_reflect_v1\",\n  \"principles\": [",
            log->mode);
    for (i = 0; i < log->n_principles; i++)
        fprintf(f, "%s\"%s\"", i ? ", " : "", log->principles[i]);
    fprintf(f, "],\n  \"goals_after\": [");
    for (i = 0; i < log->n_goals; i++)
        fprintf(f, "%s\"%s\"", i ? ", " : "", log->goals_after[i]);
    fprintf(f, "]\n}\n");
    fclose(f);
}

static int reflect(Goal *picked, int n_picked, int bonsai_ok, int lane_ok,
                   int eval_veto, int freeze_seals, int busy, double d_backlog,
                   double vigilance, double frustration, double caution,
                   double patience, char last_goals[][ID_LEN], int n_last,
                   int *same_streak, Goal *out, int *n_out, ZenLog *log) {
    Goal revised[MAX_GOALS];
    int n = 0, i;
    memset(log, 0, sizeof *log);
    for (i = 0; i < n_picked && i < MAX_GOALS; i++) revised[n++] = picked[i];

    if (bonsai_ok == 0 || lane_ok == 0) {
        add_prin(log, "engaged_action: teacher/lane down — commit keep_alive immediately");
        if (!has_id(revised, n, "keep_teacher_alive")) {
            Goal g;
            memset(&g, 0, sizeof g);
            snprintf(g.id, sizeof g.id, "keep_teacher_alive");
            snprintf(g.actions, sizeof g.actions, "ensure_bonsai,ensure_lane");
            if (n < MAX_GOALS) {
                memmove(&revised[1], &revised[0], (size_t)n * sizeof(Goal));
                revised[0] = g;
                n++;
            }
        }
        snprintf(log->mode, sizeof log->mode, "commit_infra");
        *n_out = n;
        memcpy(out, revised, (size_t)n * sizeof(Goal));
        for (i = 0; i < n; i++)
            snprintf(log->goals_after[log->n_goals++], ID_LEN, "%s", revised[i].id);
        write_log(log);
        return 0;
    }

    if (eval_veto || freeze_seals) {
        Goal filtered[MAX_GOALS];
        int nf = 0;
        add_prin(log, "non_attachment: eval_veto/freeze — drop seal-heavy goals");
        for (i = 0; i < n; i++) {
            if (strcmp(revised[i].id, "peft_jtc") == 0 ||
                strcmp(revised[i].id, "structure_mine") == 0 ||
                strcmp(revised[i].id, "break_plateau") == 0)
                continue;
            filtered[nf++] = revised[i];
        }
        n = nf;
        memcpy(revised, filtered, (size_t)n * sizeof(Goal));
        if (n == 0) {
            memset(&revised[0], 0, sizeof revised[0]);
            snprintf(revised[0].id, sizeof revised[0].id, "outcome_review");
            snprintf(revised[0].actions, sizeof revised[0].actions,
                     "close_outcome,run_eval_probe");
            n = 1;
        }
        add_note(log, "dropped_seals");
    }

    {
        int match = (n == n_last);
        for (i = 0; match && i < n; i++)
            if (strcmp(revised[i].id, last_goals[i]) != 0) match = 0;
        if (match)
            (*same_streak)++;
        else
            *same_streak = 1;
    }
    if (*same_streak >= 4 && d_backlog >= 0 && !has_id(revised, n, "outcome_review")) {
        add_prin(log, "beginners_mind: attached to same goals without backlog improvement — force review");
        memset(revised, 0, sizeof revised);
        snprintf(revised[0].id, sizeof revised[0].id, "outcome_review");
        snprintf(revised[0].actions, sizeof revised[0].actions,
                 "close_outcome,run_eval_probe");
        snprintf(revised[1].id, sizeof revised[1].id, "health_hold");
        snprintf(revised[1].actions, sizeof revised[1].actions, "log_hold");
        n = 2;
        add_note(log, "attachment_break");
    }

    if ((vigilance >= 0.72 || caution >= 0.78) && busy) {
        Goal filtered[MAX_GOALS];
        int nf = 0;
        add_prin(log, "zazen: high vigilance+busy — demote heavy explore/break");
        for (i = 0; i < n; i++) {
            if (strcmp(revised[i].id, "break_plateau") == 0 ||
                strcmp(revised[i].id, "coverage_curiosity") == 0 ||
                strcmp(revised[i].id, "verified_web") == 0)
                continue;
            filtered[nf++] = revised[i];
        }
        n = nf;
        memcpy(revised, filtered, (size_t)n * sizeof(Goal));
        if (n == 0) {
            snprintf(revised[0].id, sizeof revised[0].id, "health_hold");
            n = 1;
        }
        add_note(log, "vigilance_hold");
    }

    if (frustration >= 0.65 && patience >= 0.55 && has_id(revised, n, "drain_open_gaps") &&
        *same_streak >= 3) {
        add_prin(log, "simplicity: inject thrash — insert outcome_review before more gaps");
        if (!has_id(revised, n, "outcome_review") && n < MAX_GOALS) {
            Goal g;
            memset(&g, 0, sizeof g);
            snprintf(g.id, sizeof g.id, "outcome_review");
            memmove(&revised[1], &revised[0], (size_t)n * sizeof(Goal));
            revised[0] = g;
            n++;
            if (n > 3) n = 3;
        }
        add_note(log, "frustration_review");
    }

    if (n == 0) {
        add_prin(log, "present_moment: nothing left — health_hold");
        snprintf(revised[0].id, sizeof revised[0].id, "health_hold");
        n = 1;
    }
    if (log->n_principles == 0)
        add_prin(log, "clear_see: agenda accepted — move without hesitation");

    snprintf(log->mode, sizeof log->mode, "ok");
    *n_out = n;
    memcpy(out, revised, (size_t)n * sizeof(Goal));
    for (i = 0; i < n; i++)
        snprintf(log->goals_after[log->n_goals++], ID_LEN, "%s", revised[i].id);
    write_log(log);
    (void)caution;
    return 0;
}

static int self_test(void) {
    Goal picked[1], out[MAX_GOALS];
    ZenLog log;
    char last[1][ID_LEN];
    int streak = 3, n_out = 0, i, found;
    snprintf(picked[0].id, sizeof picked[0].id, "drain_open_gaps");
    snprintf(picked[0].actions, sizeof picked[0].actions, "inject_window_gaps");
    snprintf(last[0], ID_LEN, "drain_open_gaps");
    reflect(picked, 1, 1, 1, 0, 0, 0, 1.0, 0.4, 0.7, 0.6, 0.7, last, 1, &streak, out,
            &n_out, &log);
    found = has_id(out, n_out, "outcome_review") || log.n_notes > 0;
    if (!found) {
        fprintf(stderr, "attachment reflect failed\n");
        return 1;
    }
    reflect(NULL, 0, 0, 1, 0, 0, 0, 0, 0.5, 0.5, 0.5, 0.5, NULL, 0, &streak, out, &n_out,
            &log);
    if (strcmp(out[0].id, "keep_teacher_alive") != 0 || strcmp(log.mode, "commit_infra") != 0) {
        fprintf(stderr, "infra fire failed\n");
        return 1;
    }
    snprintf(picked[0].id, sizeof picked[0].id, "peft_jtc");
    streak = 1;
    reflect(picked, 1, 1, 1, 1, 0, 0, 0, 0.5, 0.5, 0.5, 0.5, NULL, 0, &streak, out, &n_out,
            &log);
    for (i = 0; i < n_out; i++) {
        if (strcmp(out[i].id, "peft_jtc") == 0) {
            fprintf(stderr, "veto drop failed\n");
            return 1;
        }
    }
    printf("ZEN_REFLECT_SELFTEST_PASS checks=3\n");
    return 0;
}

int main(int argc, char **argv) {
    int i;
    for (i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--test") == 0) return self_test();
    }
    printf("usage: import reflect() from governor / governor_zen_reflect --test\n");
    return 0;
}
